TypeCobol/Compiler/Symbols/FunctionSymbol.cs

```diff
@@ -0,0 +1,157 @@
+        public FunctionSymbol(String name) : base(name)
+        {
+            //Override the Kind here.
+            base.Kind = Kinds.Function;
+        }
+
+        /// <summary>
+        /// Full constructor.
+        /// </summary>
+        /// <param name="name">Function's name</param>
+        /// <param name="parameters">Function's paramaetrs</param>
+        /// <param name="retVar">Function's return variable</param>
+        public FunctionSymbol(String name, List<VariableSymbol> parameters, VariableSymbol retVar)
+            : this(name, new FunctionType(parameters, retVar))
+        {            
+        }
+
+        /// <summary>
+        /// FunctionType constructor.
+        /// </summary>
+        /// <param name="name">Function's name</param>
+        /// <param name="funType">Function's type</param>
+        public FunctionSymbol(String name, FunctionType funType)
+            : base(name)
+        {
+            this.FunctionType = funType;
+        }
+
+        /// <summary>
+        /// Function parameters.
+        /// </summary>
+        public List<VariableSymbol> Parameters => FunctionType?.Parameters;
+
+        /// <summary>
+        /// variable symbol
+        /// </summary>
+        public VariableSymbol ReturnSymbol => FunctionType?.ReturnSymbol;
+
+        /// <summary>
+        /// The return type
+        /// </summary>
+        public Type ReturnType => FunctionType?.ReturnType;
+
+        /// <summary>
+        /// The Function's type.
+        /// </summary>
+        public FunctionType FunctionType
+        {
+            get => (FunctionType)base.Type;
+            set
+            {
+                //Enter Parameters in the Scope.
+                foreach (var param in value.Parameters)
+                {
+                    LinkageStorageData.EnterIfNotExist(param);
+                }
+                base.Type = value;                
+            }
+        }
+
+        /// <summary>
+        /// Func
+        /// </summary>
+        public override Type Type
+        {
+            get => base.Type;
+            set
+            {
+                //Ensure that this is the type of a function.
+                System.Diagnostics.Debug.Assert(value is FunctionType);
+                this.FunctionType = (FunctionType)value;
+            }
+        }
+
+        /// <summary>
+        /// Function are not considered of Nested Programs.
+        /// </summary>
+        public override bool IsNested => false;
+
+        /// <summary>
+        /// Get the Variable visibility mask.
+        /// </summary>
+        public override Flags VariableVisibilityMask => Flags.GLOBAL_STORAGE;
+
+        /// <summary>
+        /// Get the type visibility mask for a procedure.
+        /// </summary>
+        public override Flags TypeVisibilityMask => Flags.Private | Flags.Public;
+
+        /// <summary>
+        /// Function have their types declared in their enclosing program.
```

Comment:
Functions can also declare their own Type.
Currently they can declare private/public typedef.
Soon, it'll be limited to typedef without visibility marker. Thus the type will only be visible to the function.

Reply:
Yes, actually types are all stored in the enclosing program, this is what I mean. But when types will be declared with no visibility thus effectively, they will be stored inside the function, so we will start the lookup within the function.

Reply:
So in fact this overrided method is not necessary, it can be removed.